### PR TITLE
First draft of SOP_OpenVDB_Extrapolate.

### PR DESCRIPTION
**Pull request by apradhana**
Signed-off-by: Andre Pradhana <andre.pradhana@dreamworks.com>

This is the first draft for SOP_OpenVDB_Extrapolate, which is the Houdini-facing interface for Fast Sweeping. At this point, I only exposed three tools from Fast Sweeping:

- maskSdf

- dilateSdf

- fogToSdf

We probably want to also add the other tools, namely:

- sdfToSdf

- fogToExt

- sdfToExt

- fogToSdfAndExt

- sdfToSdfAndExt

### CODE REVIEW

File: openvdb_houdini/SOP_OpenVDB_Extrapolate.cc
Comment (jmlait) [inline openvdb_houdini/SOP_OpenVDB_Extrapolate.cc:94]: Token: ignoretiles

File: openvdb_houdini/SOP_OpenVDB_Extrapolate.cc
Comment (jmlait) [inline openvdb_houdini/SOP_OpenVDB_Extrapolate.cc:110]: Token: pattern

**Comment by apradhana** [comment]
@jmlait: I have a couple of questions about the errors generated from the auto-check.
1. Do you know what needs to be done on my side so that this PR passes through EasyCLA?
2. Do you think that some of the errors (AZP openvdb and testhou180) are generated because FastSweeping.h has not been merged to the master branch? In that case, should I also add the version of FastSweeping that works with this SOP on this branch for the time being before taking it out before the final merge?
Thank you for the help Jeff!

**Comment by jmlait** [comment]
Yep, those compile errors appear to be the missing FastSweeping.h.  I wouldn't worry about those until FastSweeping is in, then you can do an update to it to move it to latest and they should clean up.

The trailing spaces is for you to fix though!  :>

**Comment by apradhana** [comment]
Right, right. I'll also do convergence test using this SOP and send the result to Ken et al. soon.

File: openvdb_houdini/SOP_OpenVDB_Extrapolate.cc
Comment (apradhana) [inline openvdb_houdini/SOP_OpenVDB_Extrapolate.cc:288]: @jmlait: Jeff, do you have a suggestion on how to expose the functor to the users in Houdini? According to the documentation, the functor defines the Dirichlet boundary condition and its signature is [](const Vec3R &xyz)->float.

Comment (jmlait) [reply openvdb_houdini/SOP_OpenVDB_Extrapolate.cc:288]: I think the most common use case is the user provides a VDB to provide the values of the functor.  Having a parameter callback for the functor would likely be too slow, and a VDB lets you pre-populate the isosurface with your favorite functor in an efficient manner.

File: openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Extrapolate.cc
Comment (jmlait) [inline openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Extrapolate.cc:341]: elementtype#

File: openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Extrapolate.cc
Comment (jmlait) [inline openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Extrapolate.cc:351]: This isn't a name, but a group.
matchgroup#, "Match Group"
It should also show up before Type because the match group will filter out which things this multparm will generate from?

File: openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Extrapolate.cc
Comment (jmlait) [inline openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Extrapolate.cc:372]: Should be bgfoo#, no internal capitilization.

File: openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Extrapolate.cc
Comment (jmlait) [inline openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Extrapolate.cc:376]: numextvdb

File: openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Extrapolate.cc
Comment (jmlait) [inline openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Extrapolate.cc:600]: If I understand properly, we are creatring a new grid called fooExt for the extension.  (And fooSdf down below).  Why not replace the incoming grid with the extension grid?

If I have a Cd grid I want an extended Cd grid out, not an unmodified Cd grid and a CdExt grid?

File: openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Extrapolate.cc
Comment (jmlait) [inline openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Extrapolate.cc]: This help is from vdb create?   Seems out of date with what it does.

**Review by jmlait — COMMENTED** [review]
Thanks for fixing the tokens and switching to replacement.

There seem to be many failures from float - double in the CI, though?  They seem to be in FastSweeping.
@kmuseth : Line 1414 of FastSweeping multiplies the SdfType of `w` with the ExtType of acc2->get().  This causes promotion of float -> double, so when you try and extend a Float Ext grid using a Double SDF grid, I think you get these errors.

I'm not sure it is as easy as casting `w` to ExtType because the latter might be vector?

**Review by apradhana — COMMENTED** [review]
> Thanks for fixing the tokens and switching to replacement.
> 
> There seem to be many failures from float - double in the CI, though? They seem to be in FastSweeping.
> @kmuseth : Line 1414 of FastSweeping multiplies the SdfType of `w` with the ExtType of acc2->get(). This causes promotion of float -> double, so when you try and extend a Float Ext grid using a Double SDF grid, I think you get these errors.
> 
> I'm not sure it is as easy as casting `w` to ExtType because the latter might be vector?

It's resolved for clang if I'm casting the result of the whole operation of, e.g., w*(d1.v*acc2->getValue(d1(ijk))+d2.v*acc2->getValue(d2(ijk))) to be ExtValue. In the case of d1.v*acc2->getValue(d1(ijk)) being a vector, it should call the [operator * defined in Vec3.h](https://github.com/AcademySoftwareFoundation/openvdb/blob/master/openvdb/openvdb/math/Vec3.h#L485). 

However, I'm seeing that it's failing with g++ when ExtValueT = int. This is because of conversion error casting acc2->getValue(d1(ijk)) from int to float. Do you think we need a template-specialization for this particular case (i.e. ExtValueT = int)?

**Comment by jmlait** [comment]
> It's resolved for clang if I'm casting the result of the whole operation of, e.g., w*(d1.v_acc2->getValue(d1(ijk))+d2.v_acc2->getValue(d2(ijk))) to be ExtValue. In the case of d1.v*acc2->getValue(d1(ijk)) being a vector, it should call the [operator * defined in Vec3.h](https://github.com/AcademySoftwareFoundation/openvdb/blob/master/openvdb/openvdb/math/Vec3.h#L485).
> 
> However, I'm seeing that it's failing with g++ when ExtValueT = int. This is because of conversion error casting acc2->getValue(d1(ijk)) from int to float. Do you think we need a template-specialization for this particular case (i.e. ExtValueT = int)?

Yuck.. .That actually is another problem.  When extrapolating integer grids, I'd strongly recommend *NOT* to do any weighting.  Instead, we should be doing a selection based on source voxel of highest weight.

An integer grid's values usually should never be interpolated.  If we wanted interpolation, we'd be using float or double :>  Integer grids usually are storing stuff like "Slice #" or "Region ID" which is meaningless to blend.  So we probably want specialization in extrapolate, but instead extrapolation like

setValue( d1.v > d2.v ? acc2->getValue(d1(ijk)) : acc2-> getValue(d2(ijk) );

File: openvdb/openvdb/tools/FastSweeping.h
Comment (kmuseth) [reply openvdb/openvdb/tools/FastSweeping.h:1324]: one more comment (though just a suggestion). if d1, d2, and d3 and v1, v2, and v3 were components of Vec3 vector, say d and v, then a faster (since it's branchless) and more elegant solution would simply read as:
 
return v[math::MinIndex(d)];

Not sure if that's possible without constructing temporary Vec3's, which might be costly.

**Review by kmuseth — APPROVED** [review]
Did you forget to push up your changes in response to my comments? Anyway, I don't want to block you anymore (in worst case we can always go back and add my suggestions for improvement) so I've approved your PR.

**Comment by jmlait** [comment]
I note the use of hvdb::GU_PrimVDB, which is relying on the injection of GU_PrimVDB into the openvdb_houdini namespace by GU_PrimVDB.  I think we've moved away from using that qualification, at least I don't see it outside of the From_Particles code.  It showed up for me as it doesn't work on the Houdini side as we obviously have our own GU_PrimVDB and don't inject it.

But nothing blocking, I can also remove the hvdb:: prefix from GU_PrimVDB in some future sync PR.

**Comment by jmlait** [comment]
I have it working, thank you for putting this together.

It seems however that if the field to extrapolate does not line up with the surface to extrapolate from, it does nothing silently?  Should we at least warn in this case?

I was trying to extrapolate a velocity field out from a surface..
1) It seems the two must be matching transforms, which is always a bit worrying in pipelines.  I thought the velocity field could be used as an oracle for the surface values to avoid that requirement?
2) I kind of expected to be able to do an "outwards only" extrapolation, rather than interpolation as well.  Ie, leave interior velocity values unchanged?  I can do this with a successive composite operation, of course....

If we can't have unalligned fields we should get a warning or something so people track that down.  But I don't see anything to hold off this PR, this is all stuff we can add as successive polish as well.

**Pull request merged**
